src/nvim/window.c

```diff
@@ -5195,8 +5192,6 @@ void win_free(win_T *wp, tabpage_T *tp)
   // Don't execute autocommands while the window is halfway being deleted.
   block_autocmds();
 
-  set_destroy(uint32_t, &wp->w_ns_set);
```

Comment:
Internally, storing the namespaces on the window (`w_ns_set `) might still be the right approach. Does that make cleanup easier (when a window is destroyed)? Otherwise `namespace_scope_map[..]->windows` will have stale ids in it? It also seems a bit more natural.

So the interface is `nvim__ns_set(..., {wins:...})`, but internally the namespaces are stored on the window/buffer/tabpage objects. This is a compromise because we don't have the [ideal(?) interface (global-local options)](https://github.com/neovim/neovim/pull/28728#issuecomment-2117553866).